src/Native/Unix/System.Native/pal_io.cpp

```diff
@@ -360,6 +363,28 @@ extern "C" int32_t SystemNative_ReadDirR(DIR* dir, void* buffer, int32_t bufferS
 
     // 0 returned with non-null result (guaranteed to be set to entry arg) -> success
     assert(result == entry);
+#else
+    errno = 0;
+    result = readdir(dir);
+
+    //  kernel set errno -> failure
```

Comment:
What about my comment/question about instead making this and the subsequent if block into:
```C++
if (result == nullptr)
{
    *outputEntry = { }; // managed out param must be initialized
    if (errno != 0)
    {
        assert(errno == EBADF); // Invalid directory stream descriptor
        return errno;
    }
    return -1; // shim convention for end of stream
}
```
That not only addresses (maybe purely theoretical) concerns about the readdir implementation changing errno even on success, but it also avoids an extra comparison on the success path.